### ISSUE TITLE
Inconsistent number of epochs in Validation 

### ISSUE THREAD

**Issue by vidlefak** [ultralytics/ultralytics#6490]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train, Val

### Bug

When training a YOLOv8 with a validation of 157 images and a batch size of 4 (this is the only parameter not taken by default), the validation takes 35 steps and this seems strange as 35*4 equals 140. 
What happened to the 17 others images ? 

### Environment

Ultralytics YOLOv8.0.196
Python-3.8.13
Torch-1.13

### Minimal Reproducible Example

model.train(data=yaml_dataset, epochs=100, batch=4)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@vidlefak hello there! 👋

Thank you for bringing this up. The behavior you're observing during validation with YOLOv8 can be attributed to how DataLoader in PyTorch handles batches. When the number of validation images is not perfectly divisible by the batch size, DataLoader might discard the last incomplete batch by default, depending on the `drop_last` parameter used in the DataLoader setup.

In the case of YOLOv8, the validation process is indeed using batches of images to perform the validation steps. With a total of 157 images and a batch size of 4, we would ideally expect around 39.25 (which is 157 divided by 4) steps to validate all images. Since it's not possible to have a fraction of a step, the DataLoader actually takes the floor of this number, which is 39, so it can provide complete batches. In the default configuration where `drop_last` is often set to `True`, the remaining images that do not fit into a batch (in your case, the last 1 image) are not used in the validation. That is likely why you're only seeing 35 steps, with some batches potentially combining training and validation images, depending on the exact implementation and parameters used.

This is a typical behavior in deep learning training routines to maintain consistency in batch sizes throughout the entire training and validation processes.

For a more detailed explanation about this behavior and how DataLoader interacts with your dataset, please refer to our documentation at https://docs.ultralytics.com.

Should this behavior not fit your requirements, a custom DataLoader can be created with the `drop_last=False` setting to ensure that every validation image is used, but bear in mind that this last incomplete batch will have a different batch size, which may affect batch normalization and other size-dependent processes.

I hope this helps clarify the situation! If you have more questions or need further assistance, don't hesitate to ask. Happy training! 🚀

**Comment by glenn-jocher** [comment]
👋 hi, thanks for letting us know about this possible problem with YOLO 🚀. We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub with `git pull` or `pip install -U ultralytics` to ensure your problem has not already been solved in the latest code version.
- ✅ **Unmodified** – Your problem must be reproducible using official Ultrlaytics code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/ultralytics/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃
